src/App.tsx

```diff
@@ -1,61 +1,174 @@
+import { useState } from 'react';
+import { encode, decode } from 'js-base64';
+
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+
+import { ITodo } from './types/Todo';
+
+import { TodoList } from './components/TodoList';
+
 import './App.scss';
+import { IUser } from './types/User';
+
+const preperadTodos = todosFromServer.map(todo => {
+  return {
+    ...todo,
+    user: usersFromServer
+      .find(user => user.id === todo.userId) ?? null,
+  };
+});
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+interface ITodoSetting {
+  userText: string;
+  isUserTextEmpty: boolean;
+  selectedUser: IUser | null;
+  isSubmit: boolean;
+}
 
 export const App = () => {
+  const [todos, setTodos] = useState<ITodo[]>(preperadTodos);
+
+  const initialTodoSetting = {
+    userText: '',
+    isUserTextEmpty: true,
+    selectedUser: null,
+    isSubmit: false,
+  };
+
+  const [{
+    userText,
+    isUserTextEmpty,
+    selectedUser,
+    isSubmit,
+  }, setTodoSetting] = useState<ITodoSetting>(initialTodoSetting);
+
+  const showErrorForUnselectedUser = (!selectedUser && isSubmit);
+  const showErrorForEmptyInput = (isUserTextEmpty && isSubmit);
+
+  const reset = () => {
+    setTodoSetting(initialTodoSetting);
+  };
+
+  const onFormSubmitHandler = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (!userText) {
+      setTodoSetting(prev => ({
+        ...prev,
+        isUserTextEmpty: true,
+        isSubmit: true,
+      }));
+
+      return;
+    }
+
+    if (!selectedUser) {
+      setTodoSetting(prev => ({
+        ...prev,
+        isSubmit: true,
+      }));
+
+      return;
+    }
+
+    const nextId = Math.max(...todos.map(el => el.id)) + 1;
+
+    setTodos([
+      ...todos,
+      {
+        id: nextId,
+        title: userText,
+        userId: selectedUser.id,
+        completed: false,
+        user: selectedUser,
+      },
+    ]);
+
+    reset();
+  };
+
+  const onInputHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setTodoSetting(prev => ({
+      ...prev,
+      userText: event.target.value,
+    }));
+
+    if (isUserTextEmpty) {
+      setTodoSetting(prev => ({
+        ...prev,
+        isUserTextEmpty: false,
+      }));
+    }
+  };
+
+  const decodeUser = (user: string) => {
+    return JSON.parse(decode(user));
+  };
+
+  const encodeUser = (user: IUser | null) => {
+    return encode(JSON.stringify(user));
+  };
```

Comment:
It's cool that you're trying something new👍, but I think this code is unnecessary for this task.


Maybe after refactoring code will look cleaner and you can leave this solution, but I would remove it .